Makefile

```diff
@@ -35,3 +40,33 @@ debug: e2e-env
 push-to-gcr: docker
 	docker tag experiments:$(VERSION) $(IMG)
 	gcloud docker -- push $(IMG)
+
+verify-gcr-image:
+	gcloud container images describe $(IMG) >/dev/null 2>&1
+
+push-to-pypi: pydist
+	if [ -r ~/.pypirc ]; then \
+		echo "Uploading $(VERSION)" ; \
+		twine upload wheels/experiments-$(VERSION)* ; \
+	else \
+		echo "~/.pypirc not found" ; \
+	fi;
+
+verify-pypi-wheels:
+	curl -I https://pypi.org/pypi/experiments/$(VERSION)/json 2>/dev/null | head -1 | grep 200 >/dev/null
+
+# note: can't do strict `setup.py check -s` (long_description_content_type not
+# yet understood in distutils).  See:
+# https://dustingram.com/articles/2018/03/16/markdown-descriptions-on-pypi
+# https://packaging.python.org/specifications/core-metadata/#description-content-type
+valid-release:
+	@echo "Ensuring a tagged, clean checkout for release..."
+	[[ "$(VERSION)" == $(shell git describe --abbrev=0 --tags) ]]
```

Comment:
What does this do?

Reply:
basically just makes sure we don't attempt to build and deploy wheels from a code checkout that isn't a cleanly tagged release commit.  

For instance if your current checkout has unsaved modifications or you are on a commit that is two after the last tag.  In each of these cases the computed VERSION string will have additional characters and will not match the RHS (essentially the last tag) and therefore not be considered valid for deployment.